README.md
<p align='center'><img src="https://user-images.githubusercontent.com/86114549/148660875-6ba0a22e-668e-4efb-9bf9-df6935add889.png" alt="Skidra_Logo_comp" width="300">

<p align='center'>A program to generate quantivative &amp; qualitative statistics from .csv database files to Excel via a simple UI<br>
<i>(Successor of <a href='https://github.com/Criomby/hubspool'>Hubspool</a>)</i><br>
<br>

<p align='center'><img width="591" alt="Screenshot 2022-01-08 233034" src="https://user-images.githubusercontent.com/86114549/148700854-585f62be-f07f-4c67-849b-e8b7716e0049.png"></p>
<br>

## Download Info: <br>
<h3> Latest ready-to-use version (Windows only):</h3>
<h3><a href='https://github.com/Criomby/Skidra/releases/tag/v1.2'>Download Skidra v1.2</a><br></h3>
<br>
Download the .exe and execute it as an administrator.<br>
<br>
If you have issues running the program, see the manual or the FAQ below.<br>
<h3>Be reassured: It's open source and safe for you and your data!</h3><br>

Verify your download:<br>
(SHA256sum) 539a98b830001c4b6b41eac15cc8dae40d7d13b6804df1d0d3db0e3492d6c35b<br>
(To verify your download you can use the <a href='https://github.com/Criomby/AutoHasher'>AutoHasher</a>)
<br>
<br>

## Getting started:
<i>Skidra requires you to have Python installed.<br>
For more information of the initial setup see user manual below.</i><br>

<strong>Action!</strong><br>
> 
> <strong>1. Select file:</strong> Select the .csv data file you want to analyze.<br>
> <strong>2. Options:</strong> Check the desired functions (see below).<br>
> <strong>3. Process:</strong> Execute checked functions.<br>
<br>

<strong>Options</strong><br>
> <strong>Generate Excel:</strong> Saves the output into an Excel file (see below)<br>
> <strong>ALL:</strong> Executes all analyzes mentioned below<br>
> 
> <strong>Industries:</strong> How many companies are in each industry.<br>
> <strong>Leads:</strong> How many leads are in each lead category.<br>
> <strong>Leads/ind.:</strong> How many leads in lead categories per industry.<br>
> <strong>Topleads:</strong> All companies with lead status 'Qualified Lead', 'Follow-up' or 'Contract'.<br>
> <strong>Pitches:</strong> All companies with pitch data and the respective data.<br>
> <strong>Reasons:</strong> All companies with reasons data and the respective data.<br>
<br>

## User manual:
### Initial setup
When starting Skidra for the first time, you have to select the location of your Python distribution & Skidra Python script.<br>

<i><b>Notice:</b> You have to have Python installed.<br>
If you do not have Python installed, you can get it on the <a href='https://www.python.org/downloads/'>official Website</a> or the <a href='https://www.microsoft.com/store/productId/9PJPW5LDXLZ5'>Windows Store</a>.</i><br>

<img width="350" alt="Screenshot 2022-01-08 232136" src="https://user-images.githubusercontent.com/86114549/148699791-ef1705a7-9126-4e3a-bea7-ff2f916c24dc.png">
<br>

Select the Python.exe:<br>
<img width="600" alt="Screenshot 2022-01-08 232136" src="https://user-images.githubusercontent.com/86114549/148700089-20da7134-b99d-4291-804e-74c7b62a8f4c.png">
<br>
<br>
Then select the location of the Skidra Python script:<br>
<img width="350" alt="Screenshot 2022-01-08 232302" src="https://user-images.githubusercontent.com/86114549/148700220-179f8acb-a276-428b-9b66-1d1bbad5820f.png">
<br>
<img width="600" alt="Screenshot 2022-01-08 232431" src="https://user-images.githubusercontent.com/86114549/148700223-e2a8e713-e2d6-479e-82cb-e7f5184cc165.png">
<br>
<br>
After that, the program will start right up each time, unless the Python.exe or Python script is moved.<br>
<br>
<br>

### Operation

<i>Buttons and checkboxes are explained <a href='https://github.com/Criomby/Skidra#getting-started'>above</a> already, so here are some examples of operation:</i><br>
<br>
Select a .csv file: (For informations regarding file requirements see below).<br>
<br>
<img width="650" alt="Screenshot 2022-01-08 232451" src="https://user-images.githubusercontent.com/86114549/148700942-2741a9ab-5220-41b5-8226-7a920bb7257d.png">
<br>
<img width="650" alt="Screenshot 2022-01-08 232602" src="https://user-images.githubusercontent.com/86114549/148700974-7183225e-73c2-4416-8551-a4dfde5a7026.png">
<br>
<br>
If you select 'ALL' every check will be processed, alternatively you can also select checks individually.<br>

The output will be displayed in the textbox.<br>

<img width="650" alt="Screenshot 2022-01-08 233034" src="https://user-images.githubusercontent.com/86114549/148701179-8617e831-2079-481f-9ec4-9ece53d8931e.png">


If you select 'Generate Excel', the Excel file will be saved next to your source .csv file.<br>
Also, the filepath of the Excel is shown on top.<br>
<br>
<img width="181" alt="Screenshot 2022-01-03 204804" src="https://user-images.githubusercontent.com/86114549/148701167-1a83433d-37eb-4540-ba28-e5585dc437b4.png">
<br>

The Excel file will contain the following sheets (depending on your selection of options):<br>
<br>
<img width="700" alt="Screenshot 2022-01-04 101417" src="https://user-images.githubusercontent.com/86114549/148701265-6709fed5-2946-44b6-bfd6-d5ecb3d92af5.png">
<br>
Leads:<br>
<img width="644" alt="Screenshot 2022-01-03 204853" src="https://user-images.githubusercontent.com/86114549/148701344-4922457c-9037-4e6e-972a-6d7580d16978.png">
<br>
<br>
Industries:<br>
<img width="685" alt="Screenshot 2022-01-03 204914" src="https://user-images.githubusercontent.com/86114549/148701361-abc11d78-5300-4f67-bb62-905de90cc299.png">
<br>
<br>
Leads_by_industries (only excerpt of list displayed here):<br>
![Screenshot 2022-01-03 222332](https://user-images.githubusercontent.com/86114549/148701384-5c10d01f-45b3-4956-80bf-3c6dbb8d4c6c.png)
<br>
<br>
Pitches:<br>
tbd<br>
Rejection reasons:<br>
tbd<br>
Topleads:<br>
tbd<br>





<br>
<br>

## Recommendations: <br>
The program gets every industry and lead status from HubSpot without any changes to the initial setup required.<br>
Should you import data from a different tool, read below for the requirements.<br>
<br>
However, some functions can perform additional analyses through changes to the HubSpot data structure / setup.<br>
The function 'Counts' automatically gets every industry and lead status from HubSpot and counts their occurrence.<br>
Additionally, with the following lead status categories configured in HubSpot, the function puts out a custom sorted table that can be exported:<br>
<br>
<strong>(If you use different lead categories, they will still be analyzed, but sorted by size and not custom sorted as shown below)</strong><br>

<br>
<strong>Lead status:</strong>
<ul>
  <li>Cold</li>
  <li>Cold Contacted</li>
  <li>Warm</li>
  <li>Follow-up</li>
  <li>Qualified Lead</li>
  <li>Contract</li>
  <li>Rejected</li>
  <li>Ineligible</li>
</ul>
<br>
Of course you are free to change the categories for custom sort in the source code should you require different ones.<br>
<br>
<strong>Pitches:</strong><br>
For the function 'Pitches', a company information entry has to be set up in HubSpot, where the pitch information is inserted with the following structure:<br>
<br>

> (Closed / Open) YYYY.MM.DD, time, location

(again, you are free to enter whatever you like here and the function still works, the output just might not be ideal without changes to the sort command)<br>
<br>
<strong>Rejection:</strong><br>
If a company / contact person rejects your offer, the reason for the rejection must be entered under "Company Information" under "Reason for rejection / unsuitability"
within HubSpot.<br>
This category has to be manually added to the HubSpot database default configuration.<br>
<br>

<h2>FAQ:</h2><br>
Regarding questions or other inquiries message me at:<br>
<br>
<strong>philippebraum@pm.me</strong><br>
<br>
<br>
<ul><li><strong>Data import requirements for database files:</strong><br></ul></li>
The imported data has to be in the .csv file format.<br>
When exporting the database file from from HubSpot select the following options:<br>
<br>
<img width="293" alt="Screenshot 2022-01-04 103955" src="https://user-images.githubusercontent.com/86114549/148040649-c8fdf626-d23f-46d6-951c-bba83a0b6f6d.png">
<br>
<img width='450' alt='HubSpot_export_view_settings' src='https://user-images.githubusercontent.com/86114549/148040296-97d8de17-5f73-4bcc-b23b-4eccf7dc8671.png'>
<br>
<i>For files not from HubSpot:</i><br>
The file has to contain the following columns:<br>
<ul>
  <li>Name (of the company)</li>
  <li>Lead Status</li>
  <li>Create Date</li>
  <li>Industry</li>
  <li>Company owner (internally responsible person)</li>
  <li>Pitch</li>
  <li>Reason for rejection / unsuitability.</li>
</ul>
<br>
<ul><li><strong>I have an Apple computer, can I use the ready-to-use release, too?</strong><br></ul></li>
The ready-to-use-version is Windows only atm.<br>
To run the program on MacOS, download the two scripts and run them in your local Python distribution, IDE or compile them yourself.<br>
<br>
<br>
<ul><li><strong>Windows Defender doesn't let me run the program:</strong><br></ul></li>
MS Windows' built-in antivirus Windows Defender automatically blocks any kind of unknown programs by default.<br>
<br>
<b>&nbsp&nbsp&nbsp&nbsp&nbspBe reassured:</b> Everything takes place on your local machine without connection to the internet.<br>
<br>
You can still check the source code of the executable yourself if you want to: https://github.com/Criomby/hubspool/archive/refs/tags/v2.5.2.zip<br>
<br>

To run the program, follow the process below:<br>

<img src="https://user-images.githubusercontent.com/86114549/141657595-cb6240a0-5fc0-4dd0-969f-70d4a958207e.png" alt="defender_run"></a>

<img src="https://user-images.githubusercontent.com/86114549/141657566-c661a1bd-5918-43c5-b8e3-3b61f14e79e0.png" alt="defender_more_details"></a>

<br>
<ul><li><strong>Dependencies for Python distribution?</strong><br></ul></li>
See requirements.txt.<br>
<br>

<h2>Support the project:</h2><br>
<p float='left'>
<a href='https://ethereum.org/en/stablecoins/'>
<img src="https://user-images.githubusercontent.com/86114549/123052110-be243880-d402-11eb-9f0b-77df24874278.png" alt="tether-usdt-logo" height="50"></a>
<a href='https://ethereum.org/en/stablecoins/'>
<img src="https://user-images.githubusercontent.com/86114549/122908329-4a2b5700-d354-11eb-8ba9-4fa8d2c76ed6.png" alt="usdc-large" height="50"></a>
<a href='https://ethereum.org/en/stablecoins/'>
<img src="https://user-images.githubusercontent.com/86114549/122908250-35e75a00-d354-11eb-8be1-243fcecc93c6.png" alt="dai-large" height="50"></a>
<a href='https://basicattentiontoken.org/'>
<img src="https://user-images.githubusercontent.com/86114549/132904922-1921973e-13f0-40e5-a912-2180fe2b1485.png" alt="bat_token" height="50"></a>
<a href='https://ethereum.org/en/stablecoins/'>
<img src="https://user-images.githubusercontent.com/86114549/122967139-7235ad00-d38a-11eb-86e9-b6e634a5fc75.png" alt="ETH-logo-round" height="50"></a>
</p>
Wallet adress:<br> 
0xfC56bfc44E5671fD689331490D8e6Fa5B121474F<br>
<br>
<img width="116" alt="ether_wallet_qr_code" src="https://user-images.githubusercontent.com/86114549/122909208-3f24f680-d355-11eb-88b9-c49afb867a98.png"><br>
Supported currencies: USDT, USDC, DAI, BAT, ETH <br>
<br>
<br>
© 2022 Braum


changelog.txt
Windows application (GUI & pre-checks):
-----------------------------------------

1.3 (latest)
Button 'Process' indicates current app status with color: 
Green: ready to process / Orange: busy processing;

1.2
Hands over filepath to processing script to resolve possible export file name bugs;

1.1
Error messages are displayed in output textbox;
Exported filename independent of source csv filename, 
before issues with renamed files may occurr;

1.0
First production release.



Python script (data processing):
-----------------------------------------

1.2 (latest)
Exported Excel files named 'Skidra_export_{date}.xlsx';
Enabled by receiving filepath from UI app;
Resolved export file name bugs;

1.1
Check for column names at process start and return names of mising data columns in .csv;
Improved processing script efficiency in reducing the file read access attemts - 
refactoring of .read.csv code logic;

Skidra/skidra.py
# Copyright 2022 Braum
# philippebraum@pm.me
# www.philippebraum.de

# version 1.2

import pandas as pd
from pandas.api.types import CategoricalDtype
import xlsxwriter
from distutils import util
import datetime
import os
import sys


class App:

    def __init__(self):
        self.tb_width = 127  # char width of textbox in Skidra interface
        self.company_data_filepath = sys.argv[1]
        self.data_file_directory = sys.argv[2]
        self.new_filepath = self.data_file_directory + f'\Skidra_export_{datetime.date.today()}.xlsx'

        try:
            self.hubspot_org = pd.read_csv(self.company_data_filepath)
        except FileNotFoundError as e:
            print(repr(e))
            print(f'\nFile could not be found:\n{self.company_data_filepath}\n\nPlease remove any whitespace in the filename and try again.')
        
        COLUMNS = ['Company name', 'Lead Status', 'Industry', 'Company owner', 'Create Date', 'Pitch', 'Reason for rejection / unsuitability']
        self.check_column_names(COLUMNS)

        self.excel = bool(util.strtobool(sys.argv[3]))
        self.allchk = bool(util.strtobool(sys.argv[4]))
        self.ind = bool(util.strtobool(sys.argv[5]))
        self.lds = bool(util.strtobool(sys.argv[6]))
        self.ldsind = bool(util.strtobool(sys.argv[7]))
        self.top = bool(util.strtobool(sys.argv[8]))
        self.pit = bool(util.strtobool(sys.argv[9]))
        self.res = bool(util.strtobool(sys.argv[10]))

        self.process()

    def process(self):
        print('-'*self.tb_width)
        if self.excel and os.path.isfile(self.new_filepath):
            try:
                os.remove(self.new_filepath)
                print('Existing Excel file overwritten.')
                print('-'*self.tb_width)
            except PermissionError:
                print(f'Existing savefile in use by another process.\nClose the file and try again.\n({self.new_filepath})')
                print('-'*self.tb_width)
                sys.exit()
        if self.excel:
            print(f'Excel file generated and saved:\n{self.new_filepath}')
            print('-' * self.tb_width)
        if self.ind or self.allchk:
            inds = self.get_industries()
        if self.lds or self.allchk:
            leads = self.get_leads()
        if self.ldsind or self.allchk:
            ldsbyind = self.get_leadsbyindustry()
        if self.top or self.allchk:
            topleads = self.get_topleads()
        if self.pit or self.allchk:
            pitches = self.get_pitches()
        if self.res or self.allchk:
            reasons = self.get_reasons()
        if self.excel:
            with pd.ExcelWriter(self.new_filepath) as writer:
                if self.lds or self.allchk:
                    leads.reset_index(drop=True, inplace=True)
                    leads.to_excel(writer, sheet_name='Leads', index=False)
                    workbook = writer.book
                    worksheet_leads = writer.sheets['Leads']
                    chart_leads = workbook.add_chart({'type': 'bar'})
                    chart_leads.add_series({
                        'categories': '=Leads!$A$2:$A$9',
                        'values': '=Leads!$B$2:$B$9',
                        'data_labels': {'value': True},
                        'points': [
                            {'fill': {'color': '#9da7b2'}},
                            {'fill': {'color': '#5c7f90'}},
                            {'fill': {'color': '#005778'}},
                            {'fill': {'color': '#f2a007'}},
                            {'fill': {'color': '#5b7b63'}},
                            {'fill': {'color': '#8c0410'}},
                            {'fill': {'color': '#a55b75'}},
                        ]
                    })
                    chart_leads.set_x_axis({'name': 'Number of companies'})
                    chart_leads.set_y_axis({'name': 'Status', 'reverse': True})
                    chart_leads.set_legend({'none': True})
                    chart_leads.set_title({'name': 'All lead categories'})
                    worksheet_leads.insert_chart('E2', chart_leads)
                    worksheet_leads.set_column(0, 0, 15)
                if self.ind or self.allchk:
                    inds.reset_index(drop=True, inplace=True)
                    inds.to_excel(writer, sheet_name='Industries', index=False)
                    workbook = writer.book
                    worksheet_industries = writer.sheets['Industries']
                    chart_industries = workbook.add_chart({'type': 'pie'})
                    chart_industries.add_series({
                    'categories': '=Industries!$A$2:$A$16',
                    'values': '=Industries!$B$2:$B$16',
                    'data_labels': {'value': True},
                    })
                    chart_industries.set_title({'name': 'All industries'})
                    worksheet_industries.insert_chart('E2', chart_industries)
                    worksheet_industries.set_column(0, 0, 22)
                if self.ldsind or self.allchk:
                    ldsbyind.reset_index(drop=True, inplace=True)
                    ldsbyind.to_excel(writer, sheet_name='Leads_by_industries', index=False)
                    worksheet_leadsbyindustries = writer.sheets['Leads_by_industries']
                    worksheet_leadsbyindustries.set_column(0, 0, 19)
                if self.top or self.allchk:
                    topleads.reset_index(drop=True, inplace=True)
                    topleads.to_excel(writer, sheet_name='Topleads', index=False)
                    worksheet_topleads = writer.sheets['Topleads']
                    worksheet_topleads.set_column(0, 0, 34)
                    worksheet_topleads.set_column(1, 1, 14)
                    worksheet_topleads.set_column(2, 2, 18)
                    worksheet_topleads.set_column(3, 3, 20)
                if self.pit or self.allchk:
                    pitches.reset_index(drop=True, inplace=True)
                    pitches.to_excel(writer, sheet_name='Pitches', index=False)
                    worksheet_pitches = writer.sheets['Pitches']
                    worksheet_pitches.set_column(0, 0, 21)
                    worksheet_pitches.set_column(1, 1, 39)
                if self.res or self.allchk:
                    reasons.reset_index(drop=True, inplace=True)
                    reasons.to_excel(writer, sheet_name='Rejection reasons', index=False)
                    worksheet_reasons = writer.sheets['Rejection reasons']
                    worksheet_reasons.set_column(0, 0, 37)
                    worksheet_reasons.set_column(1, 1, 18)
                    worksheet_reasons.set_column(2, 2, 12)
                    worksheet_reasons.set_column(3, 3, 70)

    def check_column_names(self, columns):
        columns_file = self.hubspot_org.columns.values.tolist()
        error = False
        missing = list()
        for i in range(0, len(columns), 1):
            if columns[i] not in columns_file:
                error = True
                missing.append(columns[i])
        if error:
            print('The following column(s) and corresponding data is missing from the database file:\n')
            for i in missing:
                print(f'--- {i} ---')
            print('\nPlease add the missing data and try again.\n\nFor additional information on file requirements see GitHub.')
            sys.exit()

    def get_leads(self):
        leads = self.lead_count()
        print(f'Leads:\n\n{leads}')
        print('-'*self.tb_width)
        return leads

    def lead_count(self):
        csv1 = self.hubspot_org[['Company name', 'Lead Status', 'Industry']].groupby(
            'Lead Status').size().to_frame().reset_index()
        csv1 = csv1.rename(columns={'Lead Status': 'Description', 0: 'Count'})
        lead_cat_order = CategoricalDtype(['Cold', 'Cold Contacted', 'Warm', 'Follow-up', 'Qualified Lead',
                                           'Contract', 'Rejected', 'Ineligible'], ordered=True)
        csv1['Description'] = csv1['Description'].astype(lead_cat_order)
        leads_inorder = csv1.sort_values('Description')
        return leads_inorder

    def get_industries(self):
        industries = self.industry_count()
        print(f'Industries:\n\n{industries}')
        print('-'*self.tb_width)
        return industries

    def industry_count(self):
        csv1 = self.hubspot_org[['Company name', 'Lead Status', 'Industry']].groupby('Industry').size().to_frame().reset_index()
        csv1 = csv1.rename(columns={'Industry': 'Description', 0: 'Count'})
        cats_disordered = csv1.sort_values(by='Count', ascending=False)
        return cats_disordered

    def get_leadsbyindustry(self):
        ldsbyind = self.leadsbyindustry()
        print(f'Leads per Industry:\n\n{ldsbyind}')
        print('-'*self.tb_width)
        return ldsbyind

    def leadsbyindustry(self):
        csv1 = self.hubspot_org[['Lead Status', 'Industry']]
        total = pd.DataFrame([['', '']], columns=pd.Index(['Lead Status', 0], dtype='object'))
        empt = pd.DataFrame([['', '']], columns=pd.Index(['Lead Status', 0], dtype='object'))
        industries = csv1['Industry'].unique().tolist()
        for i in industries:
            x = csv1[csv1['Industry'] == i].groupby(
                'Lead Status').size().to_frame().reset_index().sort_values(by=[0], ascending=False)
            dfx = pd.DataFrame([[i, '']], columns=x.columns)
            total = total.append(dfx).append(x).append(empt)
        total.rename(columns={'Lead Status': 'Description', 0: 'Count'}, inplace=True)
        return total

    def get_topleads(self):
        topleads = self.topleads()
        print(f'Topleads:\n\n{topleads}')
        print('-'*self.tb_width)
        return topleads

    def topleads(self):
        csv1 = self.hubspot_org[['Company name', 'Lead Status', 'Industry', 'Company owner']]
        tanalysis_topleads_followup = csv1[csv1['Lead Status'] == 'Follow-up']
        tanalysis_topleads_qualified = tanalysis_topleads_followup.append(csv1[csv1['Lead Status'] == 'Qualified Lead'])
        tanalysis_topleads = tanalysis_topleads_qualified.append(csv1[csv1['Lead Status'] == 'Contract'])
        return tanalysis_topleads

    def get_pitches(self):
        pitches = self.pitches()
        print(f'Pitchlist:\n\n{pitches}')
        print('-'*self.tb_width)
        return pitches

    def pitches(self):
        csv1 = self.hubspot_org[['Company name', 'Pitch']]
        filtered_csv1 = csv1[csv1['Pitch'].notnull()]
        result = filtered_csv1.sort_values(by='Pitch')
        return result

    def get_reasons(self):
        reasons = self.reasons()
        print(f'Rejection reasons:\n\n{reasons}')
        print('-'*self.tb_width)
        return reasons

    def reasons(self):  # reject reasons
        csv1 = self.hubspot_org[['Company name', 'Industry', 'Lead Status', 'Reason for rejection / unsuitability']]
        filtered_csv1 = csv1[csv1['Reason for rejection / unsuitability'].notnull()]
        result = filtered_csv1.sort_values(by='Lead Status')
        return result

    def printall_org_table(self):
        csv1 = self.hubspot_org[['Company name', 'Lead Status', 'Industry', 'Create Date']]
        return csv1


if __name__ == '__main__':
    App()


Skidra/skidra_1.3.py
# Copyright 2022 Braum
# philippebraum@pm.me
# www.philippebraum.de

# version 1.3

import pandas as pd
from pandas.api.types import CategoricalDtype
import xlsxwriter
from distutils import util
import json
import datetime
import os
import sys


class App:

    def __init__(self):
        self.WIDTH = 127  # width of textbox in Skidra interface
        print(sys.argv[1])
        args = json.loads(sys.argv[1])
        print('args:', args)
        sys.exit(0)

        self.company_data_filepath = args['filepath']
        self.data_file_directory = args['fileDir']
        self.new_filepath = self.data_file_directory + f'\\Skidra_export_{datetime.date.today()}.xlsx'

        try:
            self.hubspot_org = pd.read_csv(self.company_data_filepath)
        except FileNotFoundError as e:
            print(repr(e))
            print(f'\nFile could not be found:\n{self.company_data_filepath}\n\nPlease remove any whitespace in the filename and try again.')
        
        COLUMNS = ['Company name', 'Lead Status', 'Industry', 'Company owner', 'Create Date', 'Pitch', 'Reason for rejection / unsuitability']
        self.check_column_names(COLUMNS)

        self.excel = bool(util.strtobool(args['excel']))
        self.allchk = bool(util.strtobool(args['all']))
        self.ind = bool(util.strtobool(args['ind']))
        self.lds = bool(util.strtobool(args['lds']))
        self.ldsind = bool(util.strtobool(args['ldsind']))
        self.top = bool(util.strtobool(args['top']))
        self.pit = bool(util.strtobool(args['pit']))
        self.res = bool(util.strtobool(args['res']))

        self.process()

    def process(self):
        print('-'*self.WIDTH)
        if self.excel and os.path.isfile(self.new_filepath):
            try:
                os.remove(self.new_filepath)
                print('Existing Excel file overwritten.')
                print('-'*self.WIDTH)
            except PermissionError:
                print(f'Existing savefile in use by another process.\nClose the file and try again.\n({self.new_filepath})')
                print('-'*self.WIDTH)
                sys.exit()
        if self.excel:
            print(f'Excel file generated and saved:\n{self.new_filepath}')
            print('-' * self.WIDTH)
        if self.ind or self.allchk:
            inds = self.get_industries()
        if self.lds or self.allchk:
            leads = self.get_leads()
        if self.ldsind or self.allchk:
            ldsbyind = self.get_leadsbyindustry()
        if self.top or self.allchk:
            topleads = self.get_topleads()
        if self.pit or self.allchk:
            pitches = self.get_pitches()
        if self.res or self.allchk:
            reasons = self.get_reasons()
        if self.excel:
            with pd.ExcelWriter(self.new_filepath) as writer:
                if self.lds or self.allchk:
                    leads.reset_index(drop=True, inplace=True)
                    leads.to_excel(writer, sheet_name='Leads', index=False)
                    workbook = writer.book
                    worksheet_leads = writer.sheets['Leads']
                    chart_leads = workbook.add_chart({'type': 'bar'})
                    chart_leads.add_series({
                        'categories': '=Leads!$A$2:$A$9',
                        'values': '=Leads!$B$2:$B$9',
                        'data_labels': {'value': True},
                        'points': [
                            {'fill': {'color': '#9da7b2'}},
                            {'fill': {'color': '#5c7f90'}},
                            {'fill': {'color': '#005778'}},
                            {'fill': {'color': '#f2a007'}},
                            {'fill': {'color': '#5b7b63'}},
                            {'fill': {'color': '#8c0410'}},
                            {'fill': {'color': '#a55b75'}},
                        ]
                    })
                    chart_leads.set_x_axis({'name': 'Number of companies'})
                    chart_leads.set_y_axis({'name': 'Status', 'reverse': True})
                    chart_leads.set_legend({'none': True})
                    chart_leads.set_title({'name': 'All lead categories'})
                    worksheet_leads.insert_chart('E2', chart_leads)
                    worksheet_leads.set_column(0, 0, 15)
                if self.ind or self.allchk:
                    inds.reset_index(drop=True, inplace=True)
                    inds.to_excel(writer, sheet_name='Industries', index=False)
                    workbook = writer.book
                    worksheet_industries = writer.sheets['Industries']
                    chart_industries = workbook.add_chart({'type': 'pie'})
                    chart_industries.add_series({
                    'categories': '=Industries!$A$2:$A$16',
                    'values': '=Industries!$B$2:$B$16',
                    'data_labels': {'value': True},
                    })
                    chart_industries.set_title({'name': 'All industries'})
                    worksheet_industries.insert_chart('E2', chart_industries)
                    worksheet_industries.set_column(0, 0, 22)
                if self.ldsind or self.allchk:
                    ldsbyind.reset_index(drop=True, inplace=True)
                    ldsbyind.to_excel(writer, sheet_name='Leads_by_industries', index=False)
                    worksheet_leadsbyindustries = writer.sheets['Leads_by_industries']
                    worksheet_leadsbyindustries.set_column(0, 0, 19)
                if self.top or self.allchk:
                    topleads.reset_index(drop=True, inplace=True)
                    topleads.to_excel(writer, sheet_name='Topleads', index=False)
                    worksheet_topleads = writer.sheets['Topleads']
                    worksheet_topleads.set_column(0, 0, 34)
                    worksheet_topleads.set_column(1, 1, 14)
                    worksheet_topleads.set_column(2, 2, 18)
                    worksheet_topleads.set_column(3, 3, 20)
                if self.pit or self.allchk:
                    pitches.reset_index(drop=True, inplace=True)
                    pitches.to_excel(writer, sheet_name='Pitches', index=False)
                    worksheet_pitches = writer.sheets['Pitches']
                    worksheet_pitches.set_column(0, 0, 21)
                    worksheet_pitches.set_column(1, 1, 39)
                if self.res or self.allchk:
                    reasons.reset_index(drop=True, inplace=True)
                    reasons.to_excel(writer, sheet_name='Rejection reasons', index=False)
                    worksheet_reasons = writer.sheets['Rejection reasons']
                    worksheet_reasons.set_column(0, 0, 37)
                    worksheet_reasons.set_column(1, 1, 18)
                    worksheet_reasons.set_column(2, 2, 12)
                    worksheet_reasons.set_column(3, 3, 70)

    def check_column_names(self, columns):
        columns_file = self.hubspot_org.columns.values.tolist()
        error = False
        missing = list()
        for i in range(0, len(columns), 1):
            if columns[i] not in columns_file:
                error = True
                missing.append(columns[i])
        if error:
            print('The following column(s) and corresponding data is missing from the database file:\n')
            for i in missing:
                print(f'--- {i} ---')
            print('\nPlease add the missing data and try again.\n\nFor additional information on file requirements see GitHub.')
            sys.exit()        

    def get_leads(self):
        leads = self.lead_count()
        print(f'Leads:\n\n{leads}')
        print('-'*self.WIDTH)
        return leads

    def lead_count(self):
        csv1 = self.hubspot_org[['Company name', 'Lead Status', 'Industry']].groupby(
            'Lead Status').size().to_frame().reset_index()
        csv1 = csv1.rename(columns={'Lead Status': 'Description', 0: 'Count'})
        lead_cat_order = CategoricalDtype(['Cold', 'Cold Contacted', 'Warm', 'Follow-up', 'Qualified Lead',
                                           'Contract', 'Rejected', 'Ineligible'], ordered=True)
        csv1['Description'] = csv1['Description'].astype(lead_cat_order)
        leads_inorder = csv1.sort_values('Description')
        return leads_inorder

    def get_industries(self):
        industries = self.industry_count()
        print(f'Industries:\n\n{industries}')
        print('-'*self.WIDTH)
        return industries

    def industry_count(self):
        csv1 = self.hubspot_org[['Company name', 'Lead Status', 'Industry']].groupby('Industry').size().to_frame().reset_index()
        csv1 = csv1.rename(columns={'Industry': 'Description', 0: 'Count'})
        cats_disordered = csv1.sort_values(by='Count', ascending=False)
        return cats_disordered

    def get_leadsbyindustry(self):
        ldsbyind = self.leadsbyindustry()
        print(f'Leads per Industry:\n\n{ldsbyind}')
        print('-'*self.WIDTH)
        return ldsbyind

    def leadsbyindustry(self):
        csv1 = self.hubspot_org[['Lead Status', 'Industry']]
        total = pd.DataFrame([['', '']], columns=pd.Index(['Lead Status', 0], dtype='object'))
        empt = pd.DataFrame([['', '']], columns=pd.Index(['Lead Status', 0], dtype='object'))
        industries = csv1['Industry'].unique().tolist()
        for i in industries:
            x = csv1[csv1['Industry'] == i].groupby(
                'Lead Status').size().to_frame().reset_index().sort_values(by=[0], ascending=False)
            dfx = pd.DataFrame([[i, '']], columns=x.columns)
            total = total.append(dfx).append(x).append(empt)
        total.rename(columns={'Lead Status': 'Description', 0: 'Count'}, inplace=True)
        return total

    def get_topleads(self):
        topleads = self.topleads()
        print(f'Topleads:\n\n{topleads}')
        print('-'*self.WIDTH)
        return topleads

    def topleads(self):
        csv1 = self.hubspot_org[['Company name', 'Lead Status', 'Industry', 'Company owner']]
        tanalysis_topleads_followup = csv1[csv1['Lead Status'] == 'Follow-up']
        tanalysis_topleads_qualified = tanalysis_topleads_followup.append(csv1[csv1['Lead Status'] == 'Qualified Lead'])
        tanalysis_topleads = tanalysis_topleads_qualified.append(csv1[csv1['Lead Status'] == 'Contract'])
        return tanalysis_topleads

    def get_pitches(self):
        pitches = self.pitches()
        print(f'Pitchlist:\n\n{pitches}')
        print('-'*self.WIDTH)
        return pitches

    def pitches(self):
        csv1 = self.hubspot_org[['Company name', 'Pitch']]
        filtered_csv1 = csv1[csv1['Pitch'].notnull()]
        result = filtered_csv1.sort_values(by='Pitch')
        return result

    def get_reasons(self):
        reasons = self.reasons()
        print(f'Rejection reasons:\n\n{reasons}')
        print('-'*self.WIDTH)
        return reasons

    def reasons(self):  # reject reasons
        csv1 = self.hubspot_org[['Company name', 'Industry', 'Lead Status', 'Reason for rejection / unsuitability']]
        filtered_csv1 = csv1[csv1['Reason for rejection / unsuitability'].notnull()]
        result = filtered_csv1.sort_values(by='Lead Status')
        return result

    def printall_org_table(self):
        csv1 = self.hubspot_org[['Company name', 'Lead Status', 'Industry', 'Create Date']]
        return csv1


if __name__ == '__main__':
    App()
